r/src/compute.cpp

```diff
@@ -223,4 +205,96 @@ std::shared_ptr<arrow::Table> Table__FilterChunked(
   }
   return tab;
 }
+
+template <typename T>
+std::shared_ptr<T> MaybeUnbox(const char* class_name, SEXP x) {
+  if (Rf_inherits(x, "ArrowObject") && Rf_inherits(x, class_name)) {
+    Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<T>> obj(x);
+    return static_cast<std::shared_ptr<T>>(obj);
+  }
+  return nullptr;
+}
+
+arrow::Datum to_datum(SEXP x) {
+  if (auto array = MaybeUnbox<arrow::Array>("Array", x)) {
+    return array;
+  }
+
+  if (auto chunked_array = MaybeUnbox<arrow::ChunkedArray>("ChunkedArray", x)) {
+    return chunked_array;
+  }
+
+  if (auto batch = MaybeUnbox<arrow::RecordBatch>("RecordBatch", x)) {
+    return batch;
+  }
+
+  if (auto table = MaybeUnbox<arrow::Table>("Table", x)) {
+    return table;
+  }
+
+  if (auto scalar = MaybeUnbox<arrow::Scalar>("Scalar", x)) {
+    return scalar;
+  }
+
+  // This assumes that R objects have already been converted to Arrow objects;
+  // that seems right but should we do the wrapping here too/instead?
+  Rcpp::stop("to_datum: Not implemented for type %s", Rf_type2char(TYPEOF(x)));
+}
+
+SEXP from_datum(arrow::Datum datum) {
+  switch (datum.kind()) {
+    case arrow::Datum::SCALAR:
+      return Rcpp::wrap(datum.scalar());
+
+    case arrow::Datum::ARRAY:
+      return Rcpp::wrap(datum.make_array());
+
+    case arrow::Datum::CHUNKED_ARRAY:
+      return Rcpp::wrap(datum.chunked_array());
+
+    case arrow::Datum::RECORD_BATCH:
+      return Rcpp::wrap(datum.record_batch());
+
+    case arrow::Datum::TABLE:
+      return Rcpp::wrap(datum.table());
+
+    default:
+      break;
+  }
+
+  auto str = datum.ToString();
+  Rcpp::stop("from_datum: Not implemented for Datum %s", str.c_str());
+}
+
+std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
+    std::string func_name, List_ options) {
+  if (func_name == "filter") {
```

Comment:
Nit: might want to put these function names in constants or defines (doesn't need to be done this patch)